doc/15-event-listener.md

```diff
@@ -0,0 +1,180 @@
+# Event Listening in PHP
+
+Implementation for event listening using custom events and listeners. The primary purpose is to allow custom logic to be executed at various points in the application lifecycle.
+
+## Table of Contents
+
+- [Listener Implementation](#listener-implementation)
+- [Creating and Registering Custom User Events](#creating-and-registering-custom-user-events)
+- [Provided Events](#provided-events)
+
+---
+
+## Listener Implementation
+
+The listener is implemented using the `AsListener` attribute to specify the event it listens to and the priority.
+
+```php
+#[AsListener(event: AfterApplicationInitializationEvent::class, priority: 1)]
+function my_event_listener(AfterApplicationInitializationEvent $event): void
+{
+    // Custom logic to execute after application initialization
+}
+```
+
+---
+
+## Creating and Registering Custom User Events
+
+In this section, we'll guide you through the process of creating and registering your own custom user event.
+
+### 1. Create a Custom Event Class
+
+Start by creating a custom event class. This class should contain the data you want to associate with the event. In this example, we have a `MyEvent` class with a public property `$name`.
+
+```php
+class MyEvent
+{
+    public function __construct(
+        public string $name,
+    ) {
+    }
+}
+```
+
+### 2. Create a Task that Dispatches the Event
+
+Next, create a task function that dispatches your custom event. Use the `dispatch_event` function to trigger the event.
+
+```php
+#[AsTask(description: 'Task that dispatches an event')]
+function my_task_that_dispatches_an_event(): void
+{
+    $name = 'John Doe';
+    $event = new MyEvent($name);
+    dispatch_event(eventName: MyEvent::class, event: $event);
+}
+```
+
+In the example above, we create an instance of `MyEvent` and dispatch it using the `dispatch_event` function. Adjust the event data and dispatch logic according to your application's requirements.
+
+### 3. Register the Event Listener
+
+To respond to your custom event, create an event listener function and use the `AsListener` attribute to specify the event it listens to.
+
+```php
+#[AsListener(event: MyEvent::class, priority: 1)]
+function my_event_listener(MyEvent $event): void
```

Comment:
I'm wondering if one would want to use a listener that has internal state, like a CountListener for example that would keep track of the number of times it's called, how would you recommend to achieve this in the current implementation ?

[Symfony allows the attribute to be put on Listener class for example](https://symfony.com/doc/current/event_dispatcher.html#defining-event-listeners-with-php-attributes) , maybe it would be worth allowing it too ?

Reply:
It's not intended to work like this, I had proposed a little while ago a more [OOP format for declaring task from method of class](https://github.com/jolicode/castor/issues/184) that goes back to the same principle as this one, but it hadn't been accepted for implementation. 
I don't know if it's possible to implement.

I suggest waiting for a reply from @lyrixx on the subject